packages/airnode-node/src/evm/fulfillments/api-calls.ts

```diff
@@ -151,6 +183,65 @@ async function testAndSubmitFulfill(
   return [[...testLogs, errorLog], testErr, null];
 }
 
+export async function estimateGasAndSubmitFulfill(
+  airnodeRrp: AirnodeRrpV0,
+  request: Request<ApiCallWithResponse>,
+  options: TransactionOptions
+): Promise<LogsErrorData<Request<ApiCallWithResponse>>> {
+  const errorMessage = requests.getErrorMessage(request);
+  if (errorMessage) {
+    return submitFail(airnodeRrp, request, errorMessage, options);
+  }
+
+  // Should not throw
+  const [estimateGasLogs, estimateGasErr, estimateGasData] = await estimateGasToFulfill(airnodeRrp, request);
+
+  if (estimateGasErr || !estimateGasData) {
+    // Make static test call to get revert string
+    const [testLogs, testErr, testData] = await testFulfill(airnodeRrp, request, options);
+
+    if (testErr || (testData && !testData.callSuccess)) {
+      const updatedRequest: Request<ApiCallWithResponse> = {
+        ...request,
+        errorMessage: testErr
+          ? `${RequestErrorMessage.FulfillTransactionFailed} with error: ${testErr.message}`
+          : RequestErrorMessage.FulfillTransactionFailed,
+      };
+      const [submitLogs, submitErr, submittedRequest] = await submitFail(
+        airnodeRrp,
+        updatedRequest,
+        testErr?.message ?? decodeRevertString(testData?.callData || '0x'),
+        options
+      );
+      return [[...estimateGasLogs, ...testLogs, ...submitLogs], submitErr, submittedRequest];
+    }
+
+    // If static test call is succesful even though gas estimation failure,
+    // it'll be related RPC issue, so submit specific reason to chain
+    const updatedRequest: Request<ApiCallWithResponse> = {
+      ...request,
+      errorMessage: estimateGasErr
+        ? `${RequestErrorMessage.GasEstimationFailed} with error: ${estimateGasErr.message}`
+        : RequestErrorMessage.GasEstimationFailed,
+    };
+    const [submitLogs, submitErr, submittedRequest] = await submitFail(
+      airnodeRrp,
+      updatedRequest,
+      estimateGasErr?.message ?? RequestErrorMessage.GasEstimationFailed,
+      options
+    );
+    return [[...estimateGasLogs, ...testLogs, ...submitLogs], submitErr, submittedRequest];
+  }
+
+  // If gas estimation is success, submit fulfillment without making static test call
+  const gasLimitNoticeLog = logger.pend('INFO', `Gas limit is set to ${estimateGasData} for Request:${request.id}...`);
```

Comment:
Minor nit, but the three trailing periods should be removed because this log reports an outcome (the gas limit), not that an action is occurring (e.g. that the gas limit is being estimated like in the DEBUG log).